tools/twix/src/configuration/keys.rs

```diff
@@ -0,0 +1,233 @@
+use std::{collections::HashMap, fmt};
+
+use eframe::egui::{Key, Modifiers};
+use serde::{
+    de::{self, Deserializer},
+    Deserialize,
+};
+use thiserror::Error;
+
+#[cfg_attr(test, derive(PartialEq))]
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("Duplicate modifier `{0}`")]
+    DuplicateModifier(String),
+    #[error("Invalid modifier `{0}`")]
+    InvalidModifier(String),
+    #[error("Invalid key `{0}`")]
+    InvalidKey(String),
+}
+
+#[derive(Debug, Deserialize, Clone, Copy, PartialEq)]
+#[serde(rename_all = "snake_case")]
+pub enum KeybindAction {
+    CloseTab,
+    DuplicateTab,
+    FocusAbove,
+    FocusAddress,
+    FocusBelow,
+    FocusLeft,
+    FocusPanel,
+    FocusRight,
+    NoOp,
+    OpenSplit,
+    Reconnect,
+}
+
+#[derive(Debug, PartialEq, Eq, Hash)]
+pub struct KeybindTrigger {
+    pub key: Key,
+    pub modifiers: Modifiers,
+}
+
+impl KeybindTrigger {
+    pub fn parse_modifier(value: &&str) -> Result<Modifiers, Error> {
+        match *value {
+            "A" => Ok(Modifiers::ALT),
+            "C" => Ok(Modifiers::CTRL),
+            "S" => Ok(Modifiers::SHIFT),
+            _ => Err(Error::InvalidModifier(String::from(*value))),
+        }
+    }
+
+    pub fn parse(v: &str) -> Result<Self, Error> {
+        let parts = v.split('-').collect::<Vec<_>>();
+
+        let Some((raw_key, raw_modifiers)) = parts.split_last() else {
+            return Err(Error::InvalidKey(v.into()));
+        };
+
+        let is_single_ascii_uppercase_letter =
+            matches!(raw_key.as_bytes(), [letter] if letter.is_ascii_uppercase());
```

Comment:
Understand what do easy because good name.
Understand how do hard since use black magic.
Unsure how make better and no make code very verbose.

Reply:
What would [grug](https://grugbrain.dev/) do in this situation? :thinking: